src/core/io/writer/h5md_core.cpp

```diff
@@ -47,562 +52,340 @@ static void backup_file(const std::string &from, const std::string &to) {
+  static constexpr auto count() { return Vector2hs{1, 1}; }
+  static auto offset(hsize_t n_time_steps, hsize_t prefix) {
+    return Vector2hs{n_time_steps, prefix};
   }
-  if (write_charge) {
-#ifdef ELECTROSTATICS
-    WriteDataset(charge, "particles/atoms/charge/value", change_extent_2d,
-                 offset_2d, count_2d);
-#endif
+};
+
+} // namespace detail
+template <size_t dim, typename Op>
+void write_td_particle_property(hsize_t prefix, hsize_t n_part_global,
+                                ParticleRange const &particles,
+                                h5xx::dataset &dataset, Op op) {
+  auto const old_extents = static_cast<h5xx::dataspace>(dataset).extents();
+  auto const extent_particle_number =
+      std::max(n_part_global, old_extents[1]) - old_extents[1];
+  extend_dataset(dataset,
+                 detail::slice_info<dim>::extent(extent_particle_number));
+  auto const count = detail::slice_info<dim>::count();
+  auto offset = detail::slice_info<dim>::offset(old_extents[0], prefix);
+  for (auto const &p : particles) {
+    h5xx::write_dataset(dataset, op(p), h5xx::slice(offset, count));
+    // advance in the particle dimension
+    offset[1] += 1;
   }
 }
 
-void File::ExtendDataset(const std::string &path,
-                         const std::vector<int> &change_extent) {
-  /* Until now the h5xx does not support dataset extending, so we
-     have to use the lower level hdf5 library functions. */
-  auto &dataset = datasets[path];
-  /* Get the current dimensions of the dataspace. */
-  hid_t ds = H5Dget_space(dataset.hid());
-  auto rank = static_cast<hsize_t>(H5Sget_simple_extent_ndims(ds));
-  std::vector<hsize_t> dims(rank), maxdims(rank);
-  H5Sget_simple_extent_dims(ds, dims.data(), maxdims.data());
-  H5Sclose(ds);
-  /* Extend the dataset for another timestep (extent = 1) */
-  for (int i = 0; i < rank; i++) {
-    dims[i] += change_extent[i];
-  }
-  H5Dset_extent(dataset.hid(), dims.data()); // extend all dims is collective
-}
-
-/* data is assumed to be three dimensional */
-template <typename T>
-void File::WriteDataset(T &data, const std::string &path,
-                        const std::vector<int> &change_extent, hsize_t *offset,
-                        hsize_t *count) {
-  ExtendDataset(path, change_extent);
-  auto &dataset = datasets[path];
-  hid_t ds = H5Dget_space(dataset.hid());
-  auto rank = static_cast<hsize_t>(H5Sget_simple_extent_ndims(ds));
-  std::vector<hsize_t> maxdims(rank);
-  for (int i = 0; i < rank; i++) {
-    maxdims[i] = H5S_UNLIMITED;
-  }
-  H5Sselect_hyperslab(ds, H5S_SELECT_SET, offset, nullptr, count, nullptr);
-  /* Create a temporary dataspace. */
-  hid_t ds_new = H5Screate_simple(rank, count, maxdims.data());
-  /* Finally write the data to the dataset. */
-  H5Dwrite(dataset.hid(), dataset.get_type(), ds_new, ds, H5P_DEFAULT,
-           data.origin());
-  H5Sclose(ds_new);
-  H5Sclose(ds);
-}
+void File::write(const ParticleRange &particles, double time, int step) {
```

Comment:
You also have to take the box geometry as a parameter here, then drop the `box_geo` global and the corresponding header include.